hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/pubsub/PubSubMessageSenderProvider.java

```diff
@@ -0,0 +1,143 @@
+    }
+
+    @Override
+    public void stop() throws Exception {
+        // TODO check that method, because it looks like it is never called across the Hermes codebase
+        publishersCache.values().parallelStream().forEach(p -> {
+            p.shutdown();
+            try {
+                p.awaitTermination(30, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                logger.error("Interrupted termination of the PubSub publisher, it may result in messages loss");
+            }
+        });
+        publishingExecutorProvider.getExecutor().shutdown();
+    }
+
+    private Publisher createPublisher(PubSubSenderTarget resolvedTarget) {
+        try {
+            return Publisher.newBuilder(resolvedTarget.getTopicName())
+                    .setEndpoint(resolvedTarget.getPubSubEndpoint())
+                    .setCredentialsProvider(credentialsProvider.getProvider())
+                    .setRetrySettings(retrySettings)
+                    .setBatchingSettings(batchingSettings)
+                    .setExecutorProvider(publishingExecutorProvider)
+                    .build();
+        } catch (IOException e) {
+            logger.error("Cannot instantiate PubSub Publisher", e);
+            return null;
+        }
+    }
+
+    private BatchingSettings batchingSettings() {
+        long requestBytesThreshold = configFactory.getLongProperty(Configs.PUBSUB_SENDER_REQUEST_BYTES_THRESHOLD);
+        long messageCountBatchSize = configFactory.getLongProperty(Configs.PUBSUB_SENDER_MESSAGE_COUNT_BATCH_SIZE);
+        Duration publishDelayThreshold = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_PUBLISH_DELAY_THRESHOLD));
+
+        return BatchingSettings.newBuilder()
+                .setElementCountThreshold(messageCountBatchSize)
+                .setRequestByteThreshold(requestBytesThreshold)
+                .setDelayThreshold(publishDelayThreshold)
+                .build();
+    }
+
+    private RetrySettings retrySettings() {
+        Duration retryDelay = Duration.ofMillis(configFactory.getLongProperty(Configs.PUBSUB_SENDER_RETRY_DELAY));
+        double retryDelayMultiplier = configFactory.getDoubleProperty(Configs.PUBSUB_SENDER_RETRY_DELAY_MULTIPLIER);
+        Duration maxRetryDelay = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_MAX_RETRY_DELAY));
+        Duration initialRpcTimeout = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_INITIAL_RPC_TIMEOUT));
+        double rpcTimeoutMultiplier = configFactory.getDoubleProperty(Configs.PUBSUB_SENDER_RPC_TIMEOUT_MULTIPLIER);
+        Duration maxRpcTimeout = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_MAX_RPC_TIMEOUT));
+        Duration totalTimeout = Duration.ofMillis(
+                configFactory.getLongProperty(Configs.PUBSUB_SENDER_TOTAL_TIMEOUT));
```

Comment:
What do you think about getting these properties from subscription? We currently do so for http endpoints: [DefaultHttpRequestFactoryProvider](https://github.com/allegro/hermes/blob/b2f06c763da6606ea01635516df7b72975955ccb/hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/DefaultHttpRequestFactoryProvider.java#L11-L12).

Reply:
sounds good :)

Reply:
Sure, I meant only `maxRpcTimeout` and `totalTimeout`. My first thought was that they map directly to `socketTimeout` and `requestTimeout` from the subscription configuration. Now I have some doubts. I can look deeper into the PubSub client code to give you more precise answer.

Reply:
They don't map directly to socketTimeout and requestTimeout. I think as for PubSub being global service, we can create the `RetrySettings` and `BatchingSettings` simply as beans, we don't need any distinction between specific subscriptions.